File: python3/bron_kerbosch1.py
# coding: utf-8

from graph import UndirectedGraph, Vertex
from reporter import Reporter
from typing import List, Set


def bron_kerbosch1(graph: UndirectedGraph, reporter: Reporter):
    '''Naive Bron-Kerbosch algorithm'''
    candidates = graph.connected_nodes()
    if candidates:
        visit(
            graph=graph,
            reporter=reporter,
            candidates=candidates,
            excluded=set(),
            clique=[])


def visit(graph: UndirectedGraph, reporter: Reporter, candidates: Set[Vertex],
          excluded: Set[Vertex], clique: List[Vertex]):
    reporter.inc_count()
    if not candidates and not excluded:
        reporter.record(clique)

    while candidates:
        v = candidates.pop()
        neighbours = graph.adjacencies[v]
        assert neighbours
        visit(
            graph=graph,
            reporter=reporter,
            candidates=candidates.intersection(neighbours),
            excluded=excluded.intersection(neighbours),
            clique=clique + [v])
        excluded.add(v)

File: go/lib/bron_kerbosch1.go
package BronKerbosch

func bronKerbosch1(graph *UndirectedGraph, reporter Reporter) {
	// Naive Bron-Kerbosch algorithm
	candidates := graph.connectedVertices()
	if candidates.IsEmpty() {
		return
	}
	excluded := make(VertexSet, len(candidates))
	bronKerbosch1visit(graph, reporter, candidates, excluded, nil)
}

func bronKerbosch1visit(graph *UndirectedGraph, reporter Reporter,
	candidates VertexSet, excluded VertexSet, clique []Vertex) {
	for !candidates.IsEmpty() {
		v := candidates.PopArbitrary()
		neighbours := graph.neighbours(v)
		neighbouringCandidates := candidates.Intersection(neighbours)
		if !neighbouringCandidates.IsEmpty() {
			neighbouringExcluded := excluded.Intersection(neighbours)
			bronKerbosch1visit(graph, reporter,
				neighbouringCandidates,
				neighbouringExcluded,
				append(clique, v))
		} else if excluded.IsDisjoint(neighbours) {
			reporter.Record(append(clique, v))
		}
		excluded.Add(v)
	}
}
